App/Course App/Course App/Views/UIKit/HomeViewController.swift

```diff
@@ -0,0 +1,147 @@
+    }
+}
+
+
+final class HomeViewController: UIViewController {
+    private enum UIConstant {
+        static let minimumLineSpacing: CGFloat = 8
+        static let minimumInteritemSpacing: CGFloat = 10
+        static let edgeVertical: CGFloat = 0
+        static let edgeHorizontal: CGFloat = 4
+        static let headerReferenceSize: CGFloat = 30
+    }
+    
+    @IBOutlet private var categoriesCollectionView: UICollectionView!
+    
+    // MARK: DataSource
+    private lazy var dataProvider = MockDataProvider()
+    typealias DataSource = UICollectionViewDiffableDataSource<SectionData, [Joke]>
+    typealias Snapshot = NSDiffableDataSourceSnapshot<SectionData, [Joke]>
+    
+    private lazy var dataSource = makeDataSource()
+    private lazy var cancellables = Set<AnyCancellable>()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setup()
+        // Do any additional setup after loading the view.
+    }
+}
+
+// MARK: - UICollectionViewDataSource
+private extension HomeViewController {
+    func readData() {
+        dataProvider.$data.sink { [weak self] data in
+            self?.applySnapshot(data: data, animatingDifferences: true)
+        }
+        .store(in: &cancellables)
+    }
+
+    func applySnapshot(data: [SectionData], animatingDifferences: Bool = true) {
+        guard dataSource.snapshot().numberOfSections == 0 else {
+            let snapshot = dataSource.snapshot()
+            dataSource.apply(snapshot, animatingDifferences: animatingDifferences)
+            return
+            // snapshot.moveItem((data.first?.jokes.first)!, afterItem: (data.first?.jokes.last)!)
+        }
+
+        var snapshot = Snapshot()
+        snapshot.appendSections(data)
+
+        data.forEach { section in
+            snapshot.appendItems([section.jokes], toSection: section)
+        }
+
+        dataSource.apply(snapshot, animatingDifferences: animatingDifferences)
+    }
+
+    func makeDataSource() -> DataSource {
+        let dataSource = DataSource(collectionView: categoriesCollectionView) { collectionView, indexPath, _ in
+            let section = self.dataSource.snapshot().sectionIdentifiers[indexPath.section]
+
+            let imageHorizontalScrollCell: HorizontalScrollingCell = collectionView.dequeueReusableCell(for: indexPath)
+            imageHorizontalScrollCell.setAndReloadData(section.jokes)
+            
+            return imageHorizontalScrollCell
+        }
+
+        dataSource.supplementaryViewProvider = { collectionView, kind, indexPath in
+            guard kind == UICollectionView.elementKindSectionHeader else {
+                return nil
+            }
+
+            let section = self.dataSource.snapshot().sectionIdentifiers[indexPath.section]
+            let labelCell: LabelCollectionViewCell = collectionView.dequeueSupplementaryView(ofKind: UICollectionView.elementKindSectionHeader, for: indexPath)
```

Comment:
Would be nicer to have views itself in SwiftUI - so UICollectionViewCell uses hosting configuration to have view in SwiftUI only the container view have in UIKit